modmesh/apputil.py

```diff
@@ -70,21 +70,32 @@ def __init__(self, name):
         # Each run of the application appends a new environment.
         environ[name] = self
 
-        # default value will be overwritten when `run_code` is called
-        self._redirectStdOutFile = "stdout.txt"
-        self._redirectStdErrFile = "stderr.txt"
+        self._stdoutFd = -1
+        self._stderrFd = -1
 
     def run_code(self, code):
-        with open(self._redirectStdOutFile , 'w') as f1:
-            with contextlib.redirect_stdout(f1):
-                with open(self._redirectStdErrFile , 'w') as f2:
-                    with contextlib.redirect_stderr(f2):
-                        try:
-                            exec(code, self.globals, self.locals)
-                        except Exception as e:
-                            print(("{}: {}".format(type(e).__name__, str(e))), file=sys.stderr)
-                            print("traceback:", file=sys.stderr)
-                            traceback.print_stack()
+        import msvcrt
+
+        oldStdoutFd = sys.stdout
+        oldStderrFd = sys.stderr
+        try:
+            if self._stdoutFd <= 0:
+                raise Exception("wrong stdout fd:", self._stdoutFd)
+            if self._stderrFd <= 0:
+                raise Exception("wrong stderr fd:", self._stderrFd)
```

Comment:
`Exception` is for catching, not for raising.  A `ValueError` would be more appropriate here:
```suggestion
            if self._stdoutFd <= 0:
                raise ValueError("wrong stdout fd:", self._stdoutFd)
            if self._stderrFd <= 0:
                raise ValueError("wrong stderr fd:", self._stderrFd)
```

Reply:
I concur.  I do not find a lot of use cases of `raise Exception`.  We should always use a specific derived exception type.

If invalid fd is error state, the checking section should be moved outside the try block.  But I guess you intend to allow invalid fd and simply skip the fd processing.  In this case, the checking section should still be moved outside the try block.  You should use a different branch to handle the flow.